src/box/vinyl.c

```diff
@@ -964,8 +971,14 @@ vinyl_index_def_change_requires_rebuild(struct index *index,
 	 * won't reveal such statements, but we may still need to
 	 * compare them to statements inserted after ALTER hence
 	 * we can't narrow field types without index rebuild.
+	 *
+	 * Sic: If secondary index key parts are extended with primary
+	 * index key parts, cmp_def (hence the sorting order) will stay
+	 * the same, but we still have to rebuild the index because
+	 * the new key_def has more parts so we can't update it in-place,
+	 * see vinyl_index_update_def().
 	 */
-	if (old_cmp_def->part_count != new_cmp_def->part_count)
+	if (old_key_def->part_count != new_key_def->part_count)
```

Comment:
I guess we need to keep the old condition too. What if we add a field present in primary index and remove a field not present in primary index. `key_def` part count will not change but `cmp_def` part count will change.

Reply:
You're right, nice catch. Fixed and added a test.

<details><summary>diff</summary>

```diff
diff --git a/src/box/vinyl.c b/src/box/vinyl.c
index 45cad9731ea6..570359af9cd3 100644
--- a/src/box/vinyl.c
+++ b/src/box/vinyl.c
@@ -978,7 +978,8 @@ vinyl_index_def_change_requires_rebuild(struct index *index,
 	 * the new key_def has more parts so we can't update it in-place,
 	 * see vinyl_index_update_def().
 	 */
-	if (old_key_def->part_count != new_key_def->part_count)
+	if (old_key_def->part_count != new_key_def->part_count ||
+	    old_cmp_def->part_count != new_cmp_def->part_count)
 		return true;
 
 	for (uint32_t i = 0; i < new_cmp_def->part_count; i++) {
diff --git a/test/vinyl-luatest/gh_10095_update_index_def_test.lua b/test/vinyl-luatest/gh_10095_update_index_def_test.lua
index b0bf6813b1da..45c93483ae9a 100644
--- a/test/vinyl-luatest/gh_10095_update_index_def_test.lua
+++ b/test/vinyl-luatest/gh_10095_update_index_def_test.lua
@@ -25,9 +25,11 @@ g.test_update_index_def = function(cg)
         local s = box.schema.create_space('test', {engine = 'vinyl'})
         s:create_index('pk')
         s:create_index('sk', {parts = {{2, 'unsigned'}}})
-        s:insert({1, 10})
-        t.assert_equals(s.index.sk:get({10}), {1, 10})
+        s:insert({1, 10, 100})
+        t.assert_equals(s.index.sk:get({10}), {1, 10, 100})
         s.index.sk:alter({parts = {{2, 'unsigned'}, {1, 'unsigned'}}})
-        t.assert_equals(s.index.sk:get({10, 1}), {1, 10})
+        t.assert_equals(s.index.sk:get({10, 1}), {1, 10, 100})
+        s.index.sk:alter({parts = {{2, 'unsigned'}, {3, 'unsigned'}}})
+        t.assert_equals(s.index.sk:get({10, 100}), {1, 10, 100})
     end)
 end
```

</details>